File: hydrograph.engine/hydrograph.engine.cascading/src/main/java/hydrograph/engine/cascading/integration/HydrographRuntime.java
/*******************************************************************************
 * Copyright 2016 Capital One Services, LLC and Bitwise, Inc.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/
package hydrograph.engine.cascading.integration;

import java.io.IOException;
import java.util.Properties;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.hive.conf.HiveConf;
import org.apache.hadoop.hive.metastore.HiveMetaStoreClient;
import org.apache.hadoop.hive.metastore.api.MetaException;
import org.apache.hadoop.hive.thrift.DelegationTokenIdentifier;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.security.UserGroupInformation;
import org.apache.hadoop.security.token.Token;
import org.apache.hadoop.util.GenericOptionsParser;
import org.apache.thrift.TException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import cascading.cascade.Cascade;
import cascading.flow.Flow;
import cascading.property.AppProps;
import hydrograph.engine.cascading.assembly.generator.AssemblyGeneratorFactory;
import hydrograph.engine.core.core.HydrographDebugInfo;
import hydrograph.engine.core.core.HydrographJob;
import hydrograph.engine.core.core.HydrographRuntimeService;
import hydrograph.engine.core.helper.JAXBTraversal;
import hydrograph.engine.core.props.PropertiesLoader;
import hydrograph.engine.execution.tracking.ComponentPipeMapping;
import hydrograph.engine.execution.tracking.listener.ExecutionTrackingListener;
import hydrograph.engine.flow.utils.FlowManipulationContext;
import hydrograph.engine.flow.utils.FlowManipulationHandler;
import hydrograph.engine.hadoop.utils.HadoopConfigProvider;
import hydrograph.engine.jaxb.commontypes.TypeProperties.Property;
import hydrograph.engine.schemapropagation.SchemaFieldHandler;
import hydrograph.engine.utilities.GeneralUtilities;
import hydrograph.engine.utilities.HiveConfigurationMapping;

@SuppressWarnings({ "rawtypes" })
public class HydrographRuntime implements HydrographRuntimeService {

	private Properties hadoopProperties = new Properties();
	private FlowBuilder flowBuilder;
	private RuntimeContext runtimeContext;
	private String[] args;
	private PropertiesLoader config;
	private FlowManipulationContext flowManipulationContext;
	private static Logger LOG = LoggerFactory.getLogger(HydrographRuntime.class);

	public void executeProcess(String[] args, HydrographJob hydrographJob)  {
		this.args = args;
		config = PropertiesLoader.getInstance();
		LOG.info("Invoking initialize on runtime service");
		initialize(config.getRuntimeServiceProperties(), this.args, hydrographJob,
				new HydrographDebugInfo(null), null, null);
		LOG.info("Preparation started");
		prepareToExecute();
		LOG.info("Preparation completed. Now starting execution");
		LOG.info("Execution Started");
		execute();
		LOG.info("Execution Complete");
		oncomplete();
	}

	public void initialize(Properties config, String[] args, HydrographJob hydrographJob,
			HydrographDebugInfo hydrographDebugInfo, String jobId, String basePath) {

		AppProps.setApplicationName(hadoopProperties, hydrographJob.getJAXBObject()
				.getName());

		hadoopProperties.putAll(config);
		SchemaFieldHandler schemaFieldHandler = new SchemaFieldHandler(hydrographJob.getJAXBObject()
				.getInputsOrOutputsOrStraightPulls());
		
		flowManipulationContext = new FlowManipulationContext(hydrographJob,
				hydrographDebugInfo,schemaFieldHandler, jobId, basePath);

		hydrographJob = FlowManipulationHandler.execute(flowManipulationContext);

		if (hydrographJob.getJAXBObject().getRuntimeProperties() != null
				&& hydrographJob.getJAXBObject().getRuntimeProperties()
						.getProperty() != null) {
			for (Property property : hydrographJob.getJAXBObject()
					.getRuntimeProperties().getProperty()) {
				hadoopProperties.put(property.getName(), property.getValue());
			}
		}

		Configuration conf = new HadoopConfigProvider(hadoopProperties)
				.getJobConf();

		JAXBTraversal traversal = new JAXBTraversal(hydrographJob.getJAXBObject());

		if (traversal.isHiveComponentPresentInFlow()) {
			try {
				obtainTokenForHiveMetastore(conf);
			} catch (TException | IOException e) {
				throw new HydrographRuntimeException(e);
			}
		}

		String[] otherArgs;
		try {
			otherArgs = new GenericOptionsParser(conf, args).getRemainingArgs();
		} catch (IOException e) {
			throw new HydrographRuntimeException(e);
		}

		String argsString = "";
		for (String arg : otherArgs) {
			argsString = argsString + " " + arg;
		}
		LOG.info("After processing arguments are:" + argsString);
		this.args = otherArgs;
		// setJar(otherArgs);

		hadoopProperties.putAll(conf.getValByRegex(".*"));

		AssemblyGeneratorFactory assemblyGeneratorFactory = new AssemblyGeneratorFactory(
				hydrographJob.getJAXBObject());

		flowBuilder = new FlowBuilder();

		runtimeContext = new RuntimeContext(hydrographJob, traversal,
				hadoopProperties, assemblyGeneratorFactory,schemaFieldHandler);

		LOG.info("Graph '"
				+ runtimeContext.getHydrographJob().getJAXBObject().getName()
				+ "' initialized successfully");
	}

	@Override
	public void prepareToExecute()  {
		flowBuilder.buildFlow(runtimeContext);

		if (GeneralUtilities.IsArgOptionPresent(args, CommandLineOptionsProcessor.OPTION_DOT_PATH)) {
			writeDotFiles();
		}

	}

	@Override
	public void execute()  {
		if (GeneralUtilities.IsArgOptionPresent(args,
				CommandLineOptionsProcessor.OPTION_NO_EXECUTION)) {
			LOG.info(CommandLineOptionsProcessor.OPTION_NO_EXECUTION
					+ " option is provided so skipping execution");
			return;
		}

		for (Cascade cascade : runtimeContext.getCascade()) {
			if (ExecutionTrackingListener.isTrackingPluginPresent()) {
				ComponentPipeMapping.generateComponentToPipeMap(runtimeContext.getFlowContext());
				ComponentPipeMapping.generateComponentToFilterMap(runtimeContext);
				ExecutionTrackingListener.addListener(cascade);
			}
			cascade.complete();
		}

	}

	@Override
	public void oncomplete() {
		flowBuilder.cleanup(runtimeContext);

	}

	public Cascade[] getFlow() {
		return runtimeContext.getCascadingFlows();
	}

	private void writeDotFiles()  {

		
		
		String basePath = CommandLineOptionsProcessor.getDotPath(args);

		if (basePath == null) {
			throw new HydrographRuntimeException(
					CommandLineOptionsProcessor.OPTION_DOT_PATH
							+ " option is provided but is not followed by path");
		}
		LOG.info("Dot files will be written under " + basePath);

		String flowDotPath = basePath + "/"
				+ runtimeContext.getHydrographJob().getJAXBObject().getName() + "/"
				+ "flow";
		String flowStepDotPath = basePath + "/"
				+ runtimeContext.getHydrographJob().getJAXBObject().getName() + "/"
				+ "flowstep";

		int phaseCounter = 0;
		for (Cascade cascadingFlow : runtimeContext.getCascadingFlows()) {
			for (Flow flows : cascadingFlow.getFlows()) {
				flows.writeDOT(flowDotPath + "_" + phaseCounter);
				flows.writeStepsDOT(flowStepDotPath + "_" + phaseCounter);
				phaseCounter++;
			}
		}
	}

	private static void obtainTokenForHiveMetastore(Configuration conf)
			throws TException, IOException {
		conf.addResource(new Path(HiveConfigurationMapping
				.getHiveConf("path_to_hive_site_xml")));
		HiveConf hiveConf = new HiveConf();
		hiveConf.addResource(conf);
		try {
			UserGroupInformation ugi = UserGroupInformation.getCurrentUser();
			HiveMetaStoreClient hiveMetaStoreClient = new HiveMetaStoreClient(
					hiveConf);

			if (UserGroupInformation.isSecurityEnabled()) {
				String metastore_uri = conf.get("hive.metastore.uris");

				LOG.trace("Metastore URI:" + metastore_uri);

				// Check for local metastore
				if (metastore_uri != null && metastore_uri.length() > 0) {
					String principal = conf
							.get("hive.metastore.kerberos.principal");
					String username = ugi.getUserName();

					if (principal != null && username != null) {
						LOG.debug("username: " + username);
						LOG.debug("principal: " + principal);

						String tokenStr;
						try {
							// Get a delegation token from the Metastore.
							tokenStr = hiveMetaStoreClient.getDelegationToken(
									username, principal);
							// LOG.debug("Token String: " + tokenStr);
						} catch (TException e) {
							LOG.error(e.getMessage(), e);
							throw new RuntimeException(e);
						}

						// Create the token from the token string.
						Token<DelegationTokenIdentifier> hmsToken = new Token<>();
						hmsToken.decodeFromUrlString(tokenStr);
						// LOG.debug("Hive Token: " + hmsToken);

						// Add the token to the credentials.
						ugi.addToken(
								new Text("hive.metastore.delegation.token"),
								hmsToken);
						LOG.trace("Added hive.metastore.delegation.token to conf.");
					} else {
						LOG.debug("Username or principal == NULL");
						LOG.debug("username= " + username);
						LOG.debug("principal= " + principal);
						throw new IllegalArgumentException(
								"username and/or principal is equal to null!");
					}

				} else {
					LOG.info("HiveMetaStore configured in local mode");
				}
			}
		} catch (IOException e) {
			LOG.error(e.getMessage(), e);
			throw new RuntimeException(e);
		} catch (MetaException e) {
			LOG.error(e.getMessage(), e);
			throw new RuntimeException(e);
		}
	}

	public class HydrographRuntimeException extends RuntimeException {

		/**
		 * 
		 */
		private static final long serialVersionUID = -7891832980227676974L;

		public HydrographRuntimeException(String msg) {
			super(msg);
		}

		public HydrographRuntimeException(Throwable e) {
			super(e);
		}
	}
	
	/**
	 * Method to kill the job
	 */
	@Override
	public void kill() {
		if (runtimeContext.getCascade() != null) {
			for (Cascade cascade : runtimeContext.getCascade()) {
				cascade.stop();
			}
		} else
			System.exit(0);
	}
}

File: hydrograph.engine/hydrograph.engine.spark/src/main/scala/hydrograph/engine/spark/flow/HydrographRuntime.scala
package hydrograph.engine.spark.flow

import java.io.IOException
import java.util.Properties

import hydrograph.engine.core.core.{HydrographJob, HydrographRuntimeService}
import hydrograph.engine.core.flowmanipulation.{FlowManipulationContext, FlowManipulationHandler}
import hydrograph.engine.core.helper.JAXBTraversal
import hydrograph.engine.core.schemapropagation.SchemaFieldHandler
import hydrograph.engine.spark.components.adapter.factory.AdapterFactory
import hydrograph.engine.spark.components.base.SparkFlow
import org.apache.hadoop.fs.{FileSystem, Path}
import org.apache.spark.sql.SparkSession
import org.slf4j.{Logger, LoggerFactory}
import org.apache.spark.SparkConf
import scala.collection.JavaConverters._
import scala.collection.mutable.ListBuffer

/**
  * Created by gurdits on 10/17/2016.
  */

class HydrographRuntime extends HydrographRuntimeService {

  private val EXECUTION_TRACKING: String = "hydrograph.execution.tracking"
  private val LOG: Logger = LoggerFactory.getLogger(classOf[HydrographRuntime])
  private var flowManipulationContext: FlowManipulationContext = null;
  private var flows: ListBuffer[SparkFlow] = null



  override def kill(): Unit = {
    LOG.info("Kill signal received")
    if (RuntimeContext.instance.sparkSession != null) {
      LOG.info("Killing Spark jobs")
      RuntimeContext.instance.sparkSession.stop()
    }
    else {
      LOG.info("No Spark jobs present to kill. Exiting code.")
      System.exit(0)
    }
  }

  override def initialize(properties: Properties, args: Array[String], hydrographJob: HydrographJob,
                          jobId: String, udfPath: String): Unit = {

    val configProperties = getSparkProperties(hydrographJob)

    val sparkSessionBuilder: SparkSession.Builder = SparkSession.builder()
      .master(properties.getProperty("hydrograph.spark.master"))
      .appName(hydrographJob.getJAXBObject.getName)
      .config(configProperties)



    val schemaFieldHandler = new SchemaFieldHandler(
      hydrographJob.getJAXBObject().getInputsOrOutputsOrStraightPulls());

    flowManipulationContext = new FlowManipulationContext(hydrographJob, args, schemaFieldHandler, jobId)

    val flowManipulationHandler = new FlowManipulationHandler

    val updatedHydrographJob = flowManipulationHandler.execute(flowManipulationContext);

    val adapterFactory = AdapterFactory(updatedHydrographJob.getJAXBObject)

    val traversal = new JAXBTraversal(updatedHydrographJob.getJAXBObject());

    val sparkSession: SparkSession = checkAndEnableHiveSupport(sparkSessionBuilder, traversal, properties).getOrCreate()

    val runtimeContext = RuntimeContext(adapterFactory, traversal, updatedHydrographJob,
      flowManipulationContext.getSchemaFieldHandler, sparkSession)



    //    val EXECUTION_TRACKING = "hydrograph.execution.tracking";

    //    val oproperties = OrderedPropertiesHelper.getOrderedProperties("RegisterPlugin.properties")
    //    val executionTrackingPluginName = oproperties.getProperty(EXECUTION_TRACKING)
    //    val trackingInstance = Class.forName(executionTrackingPluginName).newInstance()
    //    val executionTrackingListener = trackingInstance.asInstanceOf[ExecutionTrackingPlugin]
    //    executionTrackingListener.addListener(runtimeContext.sparkSession)


    //    if (getExecutionTrackingClass(EXECUTION_TRACKING) != null) {
    //     val executionTrackingListener = classLoader(getExecutionTrackingClass(EXECUTION_TRACKING)).asInstanceOf[ExecutionTrackingListener]
    //      executionTrackingListener.addListener(sparkSession)
    //    }

  }

  def checkAndEnableHiveSupport(sessionBuilder: SparkSession.Builder, traversal: JAXBTraversal, properties: Properties): SparkSession.Builder = {
    LOG.trace("In method checkAndEnableHiveSupport()")
    if (traversal.isHiveComponentPresentInFlow) {
      LOG.debug("Hive components are present in flow. Enabling Hive support in SparkSession with warehouse location "+properties.getProperty("hydrograph.hive.warehouse"))
      sessionBuilder
        .config("spark.sql.warehouse.dir", properties.getProperty("hydrograph.hive.warehouse"))
        .enableHiveSupport()
    } else {
      sessionBuilder
        .config("spark.sql.warehouse.dir", properties.getProperty("hydrograph.tmp.warehouse"))
    }
    sessionBuilder
  }

  override def prepareToExecute(): Unit = {
    LOG.info("Building spark flows")
    flows = FlowBuilder(RuntimeContext.instance).buildFlow()
    LOG.info("Spark flows built successfully")
  }

  override def execute(): Unit = {
    /*if (GeneralUtilities.IsArgOptionPresent(args, CommandLineOptionsProcessor.OPTION_NO_EXECUTION)) {
      LOG.info(CommandLineOptionsProcessor.OPTION_NO_EXECUTION + " option is provided so skipping execution")
      return
    }*/
    for (sparkFlow <- flows) {
      sparkFlow.execute()
    }
    RuntimeContext.instance.sparkSession.stop()
  }

  override def getExecutionStatus: AnyRef = {
    null
  }

  override def oncomplete(): Unit = {
    //Deleting TempPath For Debug
    if (flowManipulationContext != null && flowManipulationContext.getTmpPath != null) {
      flowManipulationContext.getTmpPath.asScala.foreach(tmpPath => {
        val fullPath: Path = new Path(tmpPath)
        // do not delete the root directory
        if (fullPath.depth != 0) {
          var fileSystem: FileSystem = null
          LOG.info("Deleting temp path:" + tmpPath)
          try {
            fileSystem = FileSystem.get(RuntimeContext.instance.sparkSession.sparkContext.hadoopConfiguration)
            fileSystem.delete(fullPath, true)
          }
          catch {
            case exception: NullPointerException => {
              throw new RuntimeException(exception)
            }
            case e: IOException => {
              throw new RuntimeException(e)
            }
          }
        }
      })
    }
  }

  def classLoader[T](className: String): T = {
    val clazz = Class.forName(className).getDeclaredConstructors
    clazz(0).setAccessible(true)
    clazz(0).newInstance().asInstanceOf[T]
  }

  def getSparkProperties(hydrographJob: HydrographJob): SparkConf = {
    val configProperties = new SparkConf()
    val runttimeProperties = hydrographJob.getJAXBObject().getRuntimeProperties
    if (runttimeProperties != null) {
      for (runtimeProperty <- runttimeProperties.getProperty.asScala) {
        configProperties.set(runtimeProperty.getName, runtimeProperty.getValue)
      }
    }
    configProperties
  }


  //  def getExecutionTrackingClass(executionTrackingKey: String): String = {
  //    var properties: OrderedProperties = new OrderedProperties
  //    try {
  //      properties = OrderedPropertiesHelper.getOrderedProperties("RegisterPlugin.properties")
  //    }
  //    catch {
  //      case e: IOException => {
  //        throw new RuntimeException("Error reading the properties file: RegisterPlugin.properties" + e)
  //      }
  //    }
  //     properties.getProperty(executionTrackingKey)
  //  }
}